include/revng/Pipeline/Step.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+#include <set>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Error.h"
+
+#include "revng/ADT/STLExtras.h"
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/Contract.h"
+#include "revng/Pipeline/Pipe.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/FunctionIntrospection.h"
+
+namespace Pipeline {
+
+/// A step is a list of pipes that must be exectuted entirelly or not at all.
+/// Furthermore a step has a set of containers associated to it as well that
+/// will contain the element used for perform the computations.
+class Step {
+private:
+  std::string Name;
+  ContainerSet Containers;
+  std::vector<PipeWrapper> Pipes;
+
+public:
+  /// Clones the Targets from the backing containers of this step
+  /// and excutes all the pipes in sequence contained by this step
+  /// and returns the transformed containers.
+  ///
+  /// The contained values stays unchanged.
+  ContainerSet cloneAndRun(Context &Ctx,
+                           const ContainersStatus &Targets,
+                           llvm::raw_ostream *OS = nullptr);
+
+  /// Returns the set of goals that are already contained in the backing
+  /// containers of this step, futhermore adds to the container ToLoad those
+  /// that were not present.
+  ContainersStatus satisfiableGoals(const ContainersStatus &RequiredGoals,
+                                    ContainersStatus &ToLoad) const;
+
+  /// Returns the predicted state of the Input containers status after the
+  /// execution of all the pipes in this step
+  ContainersStatus deduceResults(ContainersStatus Input) const;
+
+  /// Consumes the provided container and merges it into the current one
+  void mergeBack(ContainerSet RunningContainers) {
```

Comment:
* We're not merging *runners*. `mergeBackContainers`? But actually, given that we're exposing `getContainers`, I think this method is superfluous.
* Would it make sense to take rvalue reference to be more explicit?